File: api/app/lib/rts_web/channels/room_channel.ex
defmodule RtsWeb.RoomChannel do
  use RtsWeb, :channel
  alias RtsWeb.Chat
  alias RtsWeb.UserStore
  alias RtsWeb.RoomJSON
  alias Rts.Rooms
  alias Rts.Rooms.Room

  @impl true
  def join("room:lobby", payload, socket) do
    if authorized?(payload) do
      {:ok, socket}
    else
      {:error, %{reason: "unauthorized"}}
    end
  end

  @impl true
  def join("room:" <> room_id, %{"user_id" => user_id, "user_name" => user_name} = payload, socket) do
    UserStore.join_room(%{room_id: room_id, user_id: user_id, user_name: user_name})
    if authorized?(payload) do
      {:ok, socket}
    else
      {:error, %{reason: "unauthorized"}}
    end
  end

  def handle_in("join_room", %{"room_id" => room_id}, socket) do
    if room_id != nil do
      room_member_list = UserStore.get_room_member_list(room_id)
      broadcast!(socket, "on_user_join", %{list: room_member_list})
    end
    {:noreply, socket}
  end

  def handle_in("on_leave_room", %{"room_id" => room_id} = payload, socket) do
    if room_id != nil do
      list = UserStore.delete_room_member(payload)
      IO.inspect "Leave result"
      IO.inspect list
      broadcast!(socket, "on_user_join", %{list: list})
    end
    {:noreply, socket}
  end

  def handle_in("list", payload, socket) do
    list = Chat.list(payload)
    {:reply, {:ok, %{list: list}}, socket}
  end

  def handle_in("new_msg", payload, socket) do
    Chat.add(payload)
    list = Chat.list(payload)
    broadcast!(socket, "new_msg", %{list: list})
    {:noreply, socket}
  end

  def handle_in("new_user", payload, socket) do
    id = UserStore.add(payload)
    {:reply, {:ok, %{id: id}}, socket}
  end

  def handle_in("on_user_inputing", payload, socket) do
    broadcast!(socket, "on_user_inputing", payload)
    {:reply, {:ok, %{}}, socket}
  end

  def handle_in("on_user_inputend", %{"user_id" => user_id} = payload, socket) do
    broadcast!(socket, "on_user_inputend", %{user_id: user_id})
    {:reply, {:ok, %{}}, socket}
  end

  def handle_in("new_room", payload, socket) do
    with {:ok, %Room{} = raw_room} <- Rooms.create_room(payload) do
      room = room_data(raw_room)
      broadcast!(socket, "on_new_room", %{data: room})
    end
    {:noreply, socket}
  end

  def room_data(room) do
    %{
      id: room.id,
      name: room.name
    }
  end

  # Add authorization logic here as required.
  defp authorized?(_payload) do
    true
  end
end

File: frontend/app/src/api/channels/roomChannel.ts
import { Socket, Channel } from 'phoenix';
import type { AnyFunction } from '../../@types';
import { User } from '../../components/providers/Room.provider';

const SOCKET_ENDPOINT = 'ws://localhost:4000/socket';

export type RoomTopic =
  | 'on_new_room'
  | 'new_msg'
  | 'on_user_inputing'
  | 'on_user_inputend'
  | 'on_user_join';

export type RoomMessagePayload = {
  user_id: string;
  message: string;
};

export type ClientChatPayload = {
  content: string;
  userId: string;
  userName: string;
};

export type ChatPayload = {
  content: string;
  room_id: string | null;
  user_id: string;
  user_name: string;
  created_at: string;
};

export type ChatListPayload = {
  list: ChatPayload[];
};

class RoomChannel {
  private channel: Channel;
  public roomId?: string | null;
  public userId?: string;

  constructor(roomId?: string | null, user?: User) {
    const socket = new Socket(SOCKET_ENDPOINT, {});

    socket.connect();
    this.channel = socket.channel(roomId ? `room:${roomId}` : 'room:lobby', {
      user_id: user?.userId,
      user_name: user?.userName,
    });
    this.roomId = roomId;
    this.userId = user?.userId;
  }

  join(onOk: AnyFunction) {
    this.channel.join().receive('ok', onOk);
  }

  disconnect() {
    this.channel.leave();
  }

  on(topic: RoomTopic, callback: AnyFunction) {
    return this.channel.on(topic, callback);
  }

  getRoomList(callback: AnyFunction) {
    return this.channel.push('room_list', {}).receive('ok', callback);
  }

  getChatList(callback: AnyFunction) {
    return this.channel
      .push('list', {
        room_id: this.roomId || null,
      })
      .receive('ok', callback);
  }

  getRoomMemberList(roomId: string, callback: AnyFunction) {
    return this.channel
      .push('join_room', {
        room_id: roomId,
      })
      .receive('ok', callback);
  }

  createDisplayUser(userName: string, callback: AnyFunction) {
    return this.channel
      .push('new_user', { name: userName })
      .receive('ok', callback);
  }

  broadCastNewRoom(name: string) {
    return this.channel.push('new_room', { name });
  }

  broadCastNewMessage({ content, userId, userName }: ClientChatPayload) {
    const roomId = this.roomId || null;
    this.channel.push('new_msg', {
      content: content,
      room_id: roomId,
      user_id: userId,
      user_name: userName,
      created_at: Date.now(),
    });
  }

  broadCastUserInputContinue(userId: string, userName: string) {
    this.channel.push('on_user_inputing', {
      user_id: userId,
      user_name: userName,
    });
  }

  broadCastUserInputEnd(userId: string) {
    this.channel.push('on_user_inputend', { user_id: userId });
  }

  broadCastLeaveRoom(userId: string, userName: string) {
    this.channel.push('on_leave_room', {
      room_id: this.roomId,
      user_id: userId,
      user_name: userName,
    });
  }
}

export default RoomChannel;
